parity/stratum.rs

```diff
@@ -16,42 +16,97 @@
 
 //! Parity sync service
 
-use std;
 use std::sync::Arc;
+use std::sync::atomic::{AtomicBool, Ordering};
 use ethcore_stratum::{Stratum as StratumServer, PushWorkHandler, RemoteJobDispatcher, ServiceConfiguration};
-use std::thread;
 use modules::service_urls;
 use boot;
 use hypervisor::service::IpcModuleId;
-use std::net::SocketAddr;
+use hypervisor::{HYPERVISOR_IPC_URL, ControlService};
+use std::net::{SocketAddr, IpAddr};
 use std::str::FromStr;
+use nanoipc;
+use std::thread;
+use ethcore::miner::stratum::{STRATUM_SOCKET_NAME, JOB_DISPATCHER_SOCKET_NAME};
+
+pub const MODULE_ID: IpcModuleId = 8000;
 
-const STRATUM_MODULE_ID: IpcModuleId = 8000;
+#[derive(Default)]
+struct StratumControlService {
+	pub stop: Arc<AtomicBool>,
+}
+
+impl ControlService for StratumControlService {
+	fn shutdown(&self) -> bool {
+		trace!(target: "hypervisor", "Received shutdown from control service");
+		self.stop.store(true, ::std::sync::atomic::Ordering::Relaxed);
+		true
+	}
+}
 
 pub fn main() {
 	boot::setup_cli_logger("stratum");
 
 	let service_config: ServiceConfiguration = boot::payload()
 		.unwrap_or_else(|e| panic!("Fatal: error reading boot arguments ({:?})", e));
 
-	let job_dispatcher = dependency!(RemoteJobDispatcher, service_urls::MINING_JOB_DISPATCHER);
+	let job_dispatcher = dependency!(
+		RemoteJobDispatcher,
+		&service_urls::with_base(&service_config.io_path, JOB_DISPATCHER_SOCKET_NAME)
+	);
+
+	let _ = boot::main_thread();
+	let service_stop = Arc::new(AtomicBool::new(false));
 
-	let stop = boot::main_thread();
 	let server =
 		StratumServer::start(
-			&SocketAddr::from_str(&service_config.listen_addr)
-				.unwrap_or_else(|e| panic!("Fatal: invalid listen address ({:?})", e)),
+			&SocketAddr::new(
+				IpAddr::from_str(&service_config.listen_addr)
+					.unwrap_or_else(|e|
+						panic!("Fatal: invalid listen address: '{}' ({:?})", &service_config.listen_addr, e)
```

Comment:
Imho does should be errors, and there should be a single place to bootstrap everything. When there is an error it should display a nice message and gracefully exit with proper error code instead of panicking and spitting out a stack trace.